File: svm/scala/svm.scala
import breeze.linalg._

class SVM(val C: Double, val kernel: String) {

  private val w = DenseVector(2)
  private var b = 0.0

  def fit(X: DenseMatrix[Double], y: DenseVector[Int], n: Int) {
    for (i <- 0 until n) {
      val score = w dot X(i) + b
      if (y(i) * score <= 1.0) {
        for (j <- 0 until 2) {
          w(j) += C * y(i) * X(i)(j)
        }
        b += C * y(i)
      }
    }
  }

  def predict(X: DenseVector[Double]) = {
    val score = w dot X + b
    score >= 0.0 ? 1 : -1
  }
}

object SVM {

  def main(args: Array[String]): Unit = {
    val svm = new SVM(1.0, "linear")
    svm.fit(DenseMatrix([[1, 2], [3, 4]], 2, 2), DenseVector([1, -1]), 2)
    println(svm.predict(DenseVector([5, 6]))) // prints 1
  }
}

File: svm/kotlin/svm.kt
import kotlin.math.max

class SVM(val C: Double, val kernel: String) {

    val w: Array<Double> = Array(2) { 0.0 }
    var b: Double = 0.0

    fun fit(X: Array<Array<Double>>, y: Array<Int>, n: Int) {
        for (i in 0 until n) {
            val score = w.dot(X[i]) + b
            if (y[i] * score <= 1.0) {
                for (j in 0 until 2) {
                    w[j] += C * y[i] * X[i][j]
                }
                b += C * y[i]
            }
        }
    }

    fun predict(X: Array<Double>): Int {
        val score = w.dot(X) + b
        return if (score >= 0.0) 1 else -1
    }
}

fun main(args: Array<String>) {
    val svm = SVM(1.0, "linear")
    svm.fit(arrayOf(arrayOf(1.0, 2.0), arrayOf(3.0, 4.0)), arrayOf(1, -1), 2)
    println(svm.predict(arrayOf(5.0, 6.0))) // prints "1"
}
